File: shared/python/aeos_shared/models/governance.py
"""Governance, policy, risk assessment, and validation models.

Covers the Runtime Governance Layer (L5) data structures.
Satisfies Requirements 5.1–5.7, 12.1–12.5.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Enumerations
# ---------------------------------------------------------------------------


class RiskLevel(str, Enum):
    """Risk classification derived from the computed risk score.

    Thresholds per design spec:
      - low:      0.0 – 6.9  → execute immediately
      - high:     7.0 – 8.9  → trigger Approval Gate
      - critical: 9.0 – 10.0 → activate Circuit Breaker
    """

    low = "low"
    high = "high"
    critical = "critical"


class GovernanceDecision(str, Enum):
    """Outcome of a Governance Layer validation request."""

    approved = "approved"
    suspended = "suspended"   # Approval Gate triggered; awaiting operator
    halted = "halted"         # Circuit Breaker activated; permanently stopped
    rejected = "rejected"     # Permission denied or policy violation


class PolicyType(str, Enum):
    """Governance policy categories stored in the ``policies`` table."""

    risk_threshold = "risk_threshold"
    permission = "permission"
    anomaly = "anomaly"
    retention = "retention"


class ScoringMethod(str, Enum):
    """Method used to compute the risk score for an action."""

    rule_based = "rule_based"
    llm = "llm"


# ---------------------------------------------------------------------------
# Risk assessment
# ---------------------------------------------------------------------------


class RiskAssessment(BaseModel):
    """Risk score and contributing factors for a proposed action.

    Produced by the risk scoring engine (rule-based or LLM-based).
    Satisfies Requirement 5.1.
    """

    model_config = ConfigDict(from_attributes=True)

    score: float = Field(ge=0.0, le=10.0)
    factors: list[str] = Field(default_factory=list)
    scoring_method: ScoringMethod


# ---------------------------------------------------------------------------
# Policy violation
# ---------------------------------------------------------------------------


class PolicyViolation(BaseModel):
    """Describes a single policy constraint violated by a proposed action or plan.

    Returned by POST /governance/validate-plan when the plan is rejected.
    Satisfies Requirements 3.3, 5.4.
    """

    model_config = ConfigDict(from_attributes=True)

    policy_id: UUID
    policy_name: str
    violation_type: str
    description: str
    step_id: UUID | None = None


# ---------------------------------------------------------------------------
# Governance result
# ---------------------------------------------------------------------------


class GovernanceResult(BaseModel):
    """Full result of a Governance Layer validation call.

    Returned by POST /governance/validate-action and
    POST /governance/validate-plan.
    """

    model_config = ConfigDict(from_attributes=True)

    decision: GovernanceDecision
    risk_assessment: RiskAssessment | None = None
    violations: list[PolicyViolation] = Field(default_factory=list)
    escalation_id: UUID | None = None
    reason: str | None = None


# ---------------------------------------------------------------------------
# Validation result (plan-level)
# ---------------------------------------------------------------------------


class ValidationResult(BaseModel):
    """Result of validating an execution plan against governance policies.

    Returned by the Planner's internal DAG check and by the Governance Layer's
    plan validation endpoint.  Satisfies Requirements 3.3, 3.4.
    """

    model_config = ConfigDict(from_attributes=True)

    is_valid: bool
    violations: list[PolicyViolation] = Field(default_factory=list)
    reason: str | None = None


# ---------------------------------------------------------------------------
# Policy
# ---------------------------------------------------------------------------


class Policy(BaseModel):
    """A governance policy record stored in the ``policies`` table.

    Supports hot-reload: when updated, the Governance Layer reloads within
    10 seconds (Requirement 12.3).
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    version: int = Field(default=1, ge=1)
    is_active: bool = True
    policy_type: PolicyType
    config: dict
    created_by: str
    created_at: datetime
    updated_at: datetime

File: shared/typescript/types/governance.ts
/**
 * Governance, policy, and audit trail types.
 * Mirrors the Python Policy, AuditTrailEntry, and related models.
 * Satisfies Requirements 5.1–5.7, 8.5, 12.1–12.5.
 */

export type PolicyType =
  | 'risk_threshold'
  | 'permission'
  | 'anomaly'
  | 'retention';

export interface Policy {
  id: string;
  name: string;
  version: number;
  is_active: boolean;
  policy_type: PolicyType;
  config: Record<string, unknown>;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface PolicyViolation {
  policy_id: string;
  policy_name: string;
  violation_type: string;
  description: string;
  step_id: string | null;
}

export interface RiskAssessment {
  score: number;
  factors: string[];
  scoring_method: 'rule_based' | 'llm';
}

export type GovernanceDecision = 'approved' | 'suspended' | 'halted' | 'rejected';

export interface ValidationResult {
  decision: GovernanceDecision;
  risk_assessment: RiskAssessment | null;
  violations: PolicyViolation[];
  escalation_id: string | null;
  reason: string | null;
}

export interface AuditTrailEntry {
  id: number;
  event_type: string;
  timestamp: string;
  agent_identity: string;
  incident_id: string | null;
  workflow_id: string | null;
  action_description: string;
  inputs: Record<string, unknown> | null;
  outputs: Record<string, unknown> | null;
  risk_score: number | null;
  prev_entry_hash: string;
}
